### PR TITLE
DD-34 corrected canvas classes and useCanvas hook

### PR DESCRIPTION
**Pull request by YuraStork**

### CODE REVIEW

File: src/canvas_classes/square.class.ts [hunk 1]
```diff
@@ -8,26 +8,26 @@ import { Tool } from './tool.class'
...
+    this.canvas.onmouseup = this.onMouseUp.bind(this)
   }
 
   private onMouseDown(e: MouseEvent) {
     this.mouseDown = true
     this.x1 = e.offsetX
     this.y1 = e.offsetY
-    this.saved = this.canvas.current.toDataURL()
+    this.saved = this.canvas.toDataURL()
   }
```
Comment (Andrew01011997) [inline src/canvas_classes/square.class.ts:31]: How many times you have this strings in your project?

Comment (Andrew01011997) [reply src/canvas_classes/square.class.ts:31]: classes also can be extendable

File: src/canvas_classes/tool.class.ts [hunk 2]
```diff
@@ -1,23 +1,25 @@
...
+  constructor(canvas: HTMLCanvasElement, socket: SocketApp, id: string) {
     this.canvas = canvas
     this.socket = socket
     this.id = id
-    this.ctx = canvas.current.getContext('2d')
-    this.width = canvas.current.width
-    this.height = canvas.current.height
+    this.ctx = canvas.getContext('2d')
+    this.width = canvas.width
+    this.height = canvas.height
```
Comment (Andrew01011997) [inline src/canvas_classes/tool.class.ts]: Do you need store `ctx`, `width` and `heaight` when you store full `canvas` object?

File: src/hooks/useCanvas/methods/snapshot.ts [hunk 3]
```diff
@@ -1,45 +1,48 @@
...
   }
 }
 
-const pushRedo = (
-  canvas: CanvasRenderingContext2D | null,
-  snapshotList: string[],
-  snapshotIndex: number,
-  setSnapshotIndex: React.Dispatch<React.SetStateAction<number>>
-) => {
+const pushRedo: ChangeSnapshot = (canvas, snapshotList, snapshotIndex, setSnapshotIndex) => {
```
Comment (Andrew01011997) [inline src/hooks/useCanvas/methods/snapshot.ts]: do you really need callback here?

File: src/hooks/useCanvas/methods/snapshot.ts [hunk 4]
```diff
@@ -1,45 +1,48 @@
...
+  setSnapshotIndex,
+  setSnapshotList
+}: HandleSnapshot) => {
   if (snapshotList.length < 10) {
-    setSnapshotList((prev) => [...prev, canvasRef.current.toDataURL()])
+    setSnapshotList((prev) => [...prev, canvas.toDataURL()])
     setSnapshotIndex((prev) => prev + 1)
   } else {
-    setSnapshotList((prev) => [...prev.slice(1, 10), canvasRef.current.toDataURL()])
+    setSnapshotList((prev) => [...prev.slice(1, 10), canvas.toDataURL()])
```
Comment (Andrew01011997) [inline src/hooks/useCanvas/methods/snapshot.ts]: change logic to leave only one set state

File: src/types/canvas.ts [hunk 5]
```diff
@@ -3,13 +3,14 @@ import { DrawTools } from 'const/enums'
...
+  ref: React.RefObject<HTMLCanvasElement>
   tool: DrawTools
   snapshot: string | null
   setToolhandler: FunctionWithParams<DrawTools>
-  changeFillStyle: FunctionWithParams<string>
-  changeStrokeStyle: FunctionWithParams<string>
-  changeLineWidth: FunctionWithParams<number>
+  changeFillStyle: (ctx: CanvasRenderingContext2D | null, color: string) => void
+  changeStrokeStyle: (ctx: CanvasRenderingContext2D | null, color: string) => void
+  changeLineWidth: (ctx: CanvasRenderingContext2D | null, size: number) => void
```
Comment (Andrew01011997) [inline src/types/canvas.ts]: create type for this

**Review by Andrew01011997 — CHANGES_REQUESTED** [review]

File: src/canvas_classes/tool.class.ts (hunk 2 above)
Comment (YuraStork) [reply src/canvas_classes/tool.class.ts]: I left only **ctx**  not to use canvas.getContext('2d') and if(canvas.getContext('2d'))

File: src/canvas_classes/square.class.ts (hunk 1 above)
Comment (YuraStork) [reply src/canvas_classes/square.class.ts:31]: this.saved should be not to all classes

**Review by Andrew01011997 — APPROVED** [review]

**Pull request merged**